src/ray/core_worker/reference_count.h

```diff
@@ -452,7 +452,11 @@ class ReferenceCounter : public ReferenceCounterInterface,
           owned_by_us(true),
           owner_address(owner_address),
           pinned_at_raylet_id(pinned_at_raylet_id),
-          is_reconstructable(is_reconstructable) {}
+          is_reconstructable(is_reconstructable) {
+      if (pinned_at_raylet_id.has_value()) {
+        locations.emplace(pinned_at_raylet_id.value());
```

Comment:
Since these pinned location updates can happen before the Plasma-triggered location reports, I opted to eagerly add them to the object's locations. The later `AddObjectLocation` call will be a no-op since the location will already exist in the set, but will still trigger a push to subscribers, which I think is desirable. We could also trigger a push to subscribers here.